updates +win:re to wrap lines safely with regards to unicode

joemfb opened the pull request:
(by converting the `tape`s in `+tank` to (list @c) before wrapping, then back)

This is a fix for unsafe unicode handling in general, and #674 specifically.

The "printable" structure `+tank`, designed for stack traces and type printing and slated for replacement, is heavily used in printing to the console. The renderer `+re` converts `+tank` to `+wall`, optionally wrapping long lines to a specified width. But since `+tank` consists of various combinations of `+tape`, which is a linked list of utf8 bytes, it's possible for `+re` to split multi-byte unicode codepoints, resulting in `%bad-text`.

For example:

```
> (~(win re [%leaf "abc’d"]) 0 5)
[%bad-text "[60 60 34 92 92 47 97 92 92 47 34 32 34 32 32 98 34 32 34 32 32 99 34 32 34 32 32 226 34 32 34 32 32 128 34 32 34 32 32 153 34 32 34 32 32 100 34 32 34 92 92 47 32 92 92 47 34 62 62 0]"]
```

This PR is a straightforward hack to make line-wrapping safe: first we make a copy of the `+tank`, converting all the tapes within to lists of codepoints, then we wrap and render the lists of codepoints, then we convert the output back to actual tapes. It's a hack because a) it's not very efficient and b) we have to lie to the type system. But it has the advantage of working without touching all the different parts of urbit that create a `+tank`, which, again, is slated for removal.

```
> (~(win re [%leaf "abc’d"]) 0 5)
<<"abc’d">>
```

These changes are all internal to `+re`, so there shouldn't be any issues pushing this out to the live network. I still need to reproduce the exact fora post scenario from #674 and confirm that no other issues are involved.

joemfb commented:
I've tested this fix against the fora post in question (the Ford UP), and it's still crashing. But now it's because of the presence of a tab character.

I think we should just strip all control characters out of cords and/or tapes that enter from the Outside. Or we could just escape them with `+mesc`. Either way, should this just be changed specifically in the fora-post code path, or should sanitation be generally applied in existing conversions (such `+to-wain:format`, `+tuba`, `+trip` and/or others)?

cgyarvin commented:
Yes but call it something else -- cores here are two-letter, so ++ug?

I like "the Outside."  Agree on stripping control characters.  I also consider \10 a control character.  You have to be 32 or taller to go on this ride.

joemfb commented:
I've renamed `+gyp:re` to `+ug:re`, and updated it to strip all ctrl characters from the tapes in the tank before handling the unicode conversions. This now passes my local tests with some arbitrary bad tapes, as well as the original Ford UP.

I think we'll need to review the general approach to handling cords/tapes/(list @c), etc. There should be some blessed, simple path for getting untrusted text into a safe, reliable state.

(And I can't figure how how to mark the requested changes as complete on this PR)

vvisigoth commented:
@cgyarvin are Joe’s changes good to go?

cgyarvin commented:
Merged, thanks.

On Wed, May 2, 2018 at 10:59 AM, Anthony Arroyo <notifications@github.com>
wrote:

> @cgyarvin <https://github.com/cgyarvin> are Joe’s changes good to go?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/725#issuecomment-386066331>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAeCi6znAreoZ3VVCqgyPQx8p7rO7ks5tufPqgaJpZM4TkjYH>
> .
>
